File: 279_Perfect_Squares/numSquares.py
import time, math

class Solution(object):
    f = [0]
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        perfSquare = [v * v for v in xrange(1, int(math.sqrt(n)) + 1)]

        for i in xrange(len(self.f), n + 1):
            self.f.append(min(self.f[i - sqr] + 1 for sqr in perfSquare if sqr <= i))
        #print len(self.f)
        return self.f[n]

sol = Solution()
n = 9975
start = time.time()
print sol.numSquares(n)
print time.time() - start

File: 279_Perfect_Squares/numSquares.cpp
#include <iostream>
#include <vector>


using namespace std;

// 1-D DP
// space: O(n)
class Solution {
    public:
        int numSquares(int n) {
            // dp[i]: #square to construct i
            vector<int> dp(n + 1, 0);
            
            int i = 0, j, val;

            // base case
            while(i * i <= n) {
                dp[i * i] = 1;
                i++;
            }

            // general case
            for(int i = 0; i <= n; i++) {
                j = 1;
                while(i + j * j <= n) {
                    val = i + j * j;
                    dp[val] = min(dp[i] + 1, dp[val]);
                    j++;
                }
            }

            return dp[n];
        }
};
